modules/integrations/arrow-integrations-retrofit-adapter/src/main/kotlin/arrow.integrations.retrofit.adapter/CallK.kt

```diff
@@ -0,0 +1,33 @@
+package arrow.integrations.retrofit.adapter
+
+import arrow.Kind
+import arrow.effects.typeclasses.Async
+import arrow.effects.typeclasses.MonadDefer
+import arrow.typeclasses.MonadError
+import retrofit2.Call
+import retrofit2.HttpException
+import retrofit2.Response
+
+data class CallK<R>(val call: Call<R>) {
+  fun <F> async(async: Async<F>): Kind<F, R> =
+    async.async { proc -> call.enqueue(ResponseCallback(proc)) }
+
+  fun <F> defer(defer: MonadDefer<F>): Kind<F, R> =
+    defer {
+      handleResponse(call.execute())
+    }
+
+  fun <F> catch(defer: MonadError<F, Throwable>): Kind<F, R> =
+    defer.run {
+      catch {
+        handleResponse(call.execute())
+      }
+    }
+}
+
+private fun <R> handleResponse(response: Response<R>) =
+  if (response.isSuccessful) {
+    response.body() ?: throw IllegalStateException("The request returned a null body")
```

Comment:
Don't throw, use `raiseException` instead.

Reply:
Do you mean `raiseError`? I don't see a raiseException in the code. 

But, anyway, then I would need to wrap an then unwrap the result to conform with the signature:
```
invoke(f: () -> A)
```

defer is already catching any kind of Throwable. Wouldn't throw make the code simpler?

Reply:
Then, in this case `defer` should behave differently than `catch`. Defer returns a value, catch will throw.

Reply:
`catch` does not return a kinded value but `deffer` does. In whatever case, we should not throw exceptions since these integrations are used by others as an example and Arrow provides mechanisms to avoid throwing exceptions and instead of favors representing them as value. You don't need `catch` wrapping those if you just call `raiseError` instead of throwing exceptions.